### ISSUE TITLE
error with broom

### ISSUE THREAD

**Issue by minor-ls** [Torvaney/regista#32]
Hi,

I've been playing around with regista and trying to get the Dixon-Coles model to work using shot data on the Finnish Veikkausliiga, but getting stuck because of an issue with broom - as can be seen from the below reprex, it gives the error: "Error: No tidy method for objects of class dixoncoles" - this seems to be true for all broom functions, as I was using broom::augment when I first noticed the issue. I replicated the issue using exact code from your blog post (http://www.statsandsnakeoil.com/2018/06/22/dixon-coles-and-xg-together-at-last/) so I don't think it should be an issue on my end.

Thanks,

Axel
-----------------------------------------------------------------

```r
library(tidyverse)

games <-
  read_csv("https://git.io/fNmRy") %>%
  filter(season == 2017) %>%
  nest(side, xg, .key = "shots")
#> Parsed with column specification:
#> cols(
#>   match_id = col_double(),
#>   date = col_datetime(format = ""),
#>   home = col_character(),
#>   away = col_character(),
#>   hgoals = col_double(),
#>   agoals = col_double(),
#>   side = col_character(),
#>   xg = col_double(),
#>   league = col_character(),
#>   season = col_double()
#> )
#> Warning: All elements of `...` must be named.
#> Did you want `shots = c(side, xg)`?

simulate_shots <- function(xgs) {
  tibble::tibble(goals = 0:length(xgs),
                 prob  = poisbinom::dpoisbinom(0:length(xgs), xgs))
}

simulate_game <- function(shot_xgs) {
  home_xgs <- shot_xgs %>% dplyr::filter(side == "h") %>% pull(xg)
  away_xgs <- shot_xgs %>% dplyr::filter(side == "a") %>% pull(xg)
  
  home_probs <- simulate_shots(home_xgs) %>% dplyr::rename_all(function(x) paste0("h", x))
  away_probs <- simulate_shots(away_xgs) %>% dplyr::rename_all(function(x) paste0("a", x))
  
  tidyr::crossing(home_probs, away_probs) %>%
    dplyr::mutate(prob = .data$hprob * .data$aprob)
}

simulated_games <-
  games %>%
  mutate(simulated_probabilities = map(shots, simulate_game)) %>%
  select(match_id, home, away, simulated_probabilities) %>%
  unnest(cols = c(simulated_probabilities)) %>%
  filter(prob > 0.001)  # Keep the number of rows vaguely reasonable

library(regista)

# Fit a "vanilla" Dixon-Coles model (on observed goals)
fit_vanilla <- dixoncoles(
  hgoal = hgoals,
  agoal = agoals,
  hteam = home,
  ateam = away,
  data  = factor_teams(games, c("home", "away"))
)
#> Warning in model.matrix.default(~values - 1, model.frame(~values - 1), contrasts
#> = FALSE): non-list contrasts argument ignored
#> Warning in model.matrix.default(~values - 1, model.frame(~values - 1), contrasts
#> = FALSE): non-list contrasts argument ignored

#> Warning in model.matrix.default(~values - 1, model.frame(~values - 1), contrasts
#> = FALSE): non-list contrasts argument ignored

#> Warning in model.matrix.default(~values - 1, model.frame(~values - 1), contrasts
#> = FALSE): non-list contrasts argument ignored

# Fit on the simulated data, weighted by probability
fit_simulated <- dixoncoles(
  hgoal   = hgoals,
  agoal   = agoals,
  hteam   = home,
  ateam   = away,
  weights = prob,
  data    = factor_teams(simulated_games, c("home", "away")))
#> Warning in model.matrix.default(~values - 1, model.frame(~values - 1), contrasts
#> = FALSE): non-list contrasts argument ignored

#> Warning in model.matrix.default(~values - 1, model.frame(~values - 1), contrasts
#> = FALSE): non-list contrasts argument ignored

#> Warning in model.matrix.default(~values - 1, model.frame(~values - 1), contrasts
#> = FALSE): non-list contrasts argument ignored

#> Warning in model.matrix.default(~values - 1, model.frame(~values - 1), contrasts
#> = FALSE): non-list contrasts argument ignored

estimates <-
  inner_join(
    broom::tidy(fit_vanilla),
    broom::tidy(fit_simulated),
    by = c("parameter", "team"),
    suffix = c("_vanilla", "_xg")
  ) %>%
  mutate(value_vanilla = exp(value_vanilla),
         value_xg      = exp(value_xg))
#> Error: No tidy method for objects of class dixoncoles
```
Created on 2020-08-20 by the reprex package (v0.3.0)

**Comment by Torvaney** [comment]
Thanks @minor-ls 

Could you post the output of `devtools::session_info()`, please? I can't replicate the issue, so I'm wondering if we're using different package versions.

**Comment by minor-ls** [comment]
Here's the session info:

setting  value                       
 version  R version 4.0.2 (2020-06-22)
 os       Windows 10 x64              
 system   x86_64, mingw32             
 ui       RStudio                     
 language (EN)                        
 collate  English_Finland.1252        
 ctype    English_Finland.1252        
 tz       Europe/Helsinki             
 date     2020-08-20                  

- Packages ---------------------------------------------------------------------------------------------------------------
 ! package     * version    date       lib source                           
   assertthat    0.2.1      2019-03-21 [1] CRAN (R 4.0.2)                   
   backports     1.1.7      2020-05-13 [1] CRAN (R 4.0.0)                   
   blob          1.2.1      2020-01-20 [1] CRAN (R 4.0.2)                   
   broom         0.7.0      2020-07-09 [1] CRAN (R 4.0.2)                   
   callr         3.4.3      2020-03-28 [1] CRAN (R 4.0.2)                   
   cellranger    1.1.0      2016-07-27 [1] CRAN (R 4.0.2)                   
   cli           2.0.2      2020-02-28 [1] CRAN (R 4.0.2)                   
   clipr         0.7.0      2019-07-23 [1] CRAN (R 4.0.2)                   
   colorspace    1.4-1      2019-03-18 [1] CRAN (R 4.0.2)                   
   crayon        1.3.4      2017-09-16 [1] CRAN (R 4.0.2)                   
   curl          4.3        2019-12-02 [1] CRAN (R 4.0.2)                   
   DBI           1.1.0      2019-12-15 [1] CRAN (R 4.0.2)                   
   dbplyr        1.4.4      2020-05-27 [1] CRAN (R 4.0.2)                   
   desc          1.2.0      2018-05-01 [1] CRAN (R 4.0.2)                   
   devtools      2.3.1      2020-07-21 [1] CRAN (R 4.0.2)                   
   digest        0.6.25     2020-02-23 [1] CRAN (R 4.0.2)                   
   dplyr       * 1.0.0      2020-05-29 [1] CRAN (R 4.0.2)                   
   ellipsis      0.3.1      2020-05-15 [1] CRAN (R 4.0.2)                   
   evaluate      0.14       2019-05-28 [1] CRAN (R 4.0.2)                   
   fansi         0.4.1      2020-01-08 [1] CRAN (R 4.0.2)                   
   forcats     * 0.5.0      2020-03-01 [1] CRAN (R 4.0.2)                   
   fs            1.5.0      2020-07-31 [1] CRAN (R 4.0.2)                   
   generics      0.0.2      2018-11-29 [1] CRAN (R 4.0.2)                   
   ggplot2     * 3.3.2      2020-06-19 [1] CRAN (R 4.0.2)                   
   glue          1.4.1      2020-05-13 [1] CRAN (R 4.0.2)                   
   gtable        0.3.0      2019-03-25 [1] CRAN (R 4.0.2)                   
   haven         2.3.1      2020-06-01 [1] CRAN (R 4.0.2)                   
   hms           0.5.3      2020-01-08 [1] CRAN (R 4.0.2)                   
   htmltools     0.5.0      2020-06-16 [1] CRAN (R 4.0.2)                   
   httr          1.4.2      2020-07-20 [1] CRAN (R 4.0.2)                   
   jsonlite      1.7.0      2020-06-25 [1] CRAN (R 4.0.2)                   
   knitr         1.29       2020-06-23 [1] CRAN (R 4.0.2)                   
   lazyeval      0.2.2      2019-03-15 [1] CRAN (R 4.0.2)                   
   lifecycle     0.2.0      2020-03-06 [1] CRAN (R 4.0.2)                   
   lubridate   * 1.7.9      2020-06-08 [1] CRAN (R 4.0.2)                   
   magrittr      1.5        2014-11-22 [1] CRAN (R 4.0.2)                   
   memoise       1.1.0      2017-04-21 [1] CRAN (R 4.0.2)                   
   modelr        0.1.8      2020-05-19 [1] CRAN (R 4.0.2)                   
   munsell       0.5.0      2018-06-12 [1] CRAN (R 4.0.2)                   
 V pillar        1.4.4      2020-07-10 [1] CRAN (R 4.0.2)                   
   pkgbuild      1.1.0      2020-07-13 [1] CRAN (R 4.0.2)                   
   pkgconfig     2.0.3      2019-09-22 [1] CRAN (R 4.0.2)                   
   pkgload       1.1.0      2020-05-29 [1] CRAN (R 4.0.2)                   
   poisbinom     1.0.1      2017-05-19 [1] CRAN (R 4.0.2)                   
   prettyunits   1.1.1      2020-01-24 [1] CRAN (R 4.0.2)                   
   processx      3.4.3      2020-07-05 [1] CRAN (R 4.0.2)                   
   ps            1.3.4      2020-08-11 [1] CRAN (R 4.0.2)                   
   purrr       * 0.3.4      2020-04-17 [1] CRAN (R 4.0.2)                   
   R6            2.4.1      2019-11-12 [1] CRAN (R 4.0.2)                   
   Rcpp          1.0.4.6    2020-04-09 [1] CRAN (R 4.0.2)                   
   readr       * 1.3.1      2018-12-21 [1] CRAN (R 4.0.2)                   
   readxl        1.3.1      2019-03-13 [1] CRAN (R 4.0.2)                   
   regista     * 0.4.1.9000 2020-07-04 [1] Github (torvaney/regista@7e6288a)
   remotes       2.2.0      2020-07-21 [1] CRAN (R 4.0.2)                   
   reprex      * 0.3.0      2019-05-16 [1] CRAN (R 4.0.2)                   
   rlang         0.4.6      2020-05-02 [1] CRAN (R 4.0.2)                   
   rmarkdown     2.3        2020-06-18 [1] CRAN (R 4.0.2)                   
   rprojroot     1.3-2      2018-01-03 [1] CRAN (R 4.0.2)                   
   rstudioapi    0.11       2020-02-07 [1] CRAN (R 4.0.2)                   
   rvest         0.3.6      2020-07-25 [1] CRAN (R 4.0.2)                   
   scales        1.1.1      2020-05-11 [1] CRAN (R 4.0.2)                   
   sessioninfo   1.1.1      2018-11-05 [1] CRAN (R 4.0.2)                   
   stringi       1.4.6      2020-02-17 [1] CRAN (R 4.0.0)                   
   stringr     * 1.4.0      2019-02-10 [1] CRAN (R 4.0.2)                   
   testthat      2.3.2      2020-03-02 [1] CRAN (R 4.0.2)                   
   tibble      * 3.0.1      2020-04-20 [1] CRAN (R 4.0.2)                   
   tidyr       * 1.1.0      2020-05-20 [1] CRAN (R 4.0.2)                   
   tidyselect    1.1.0      2020-05-11 [1] CRAN (R 4.0.2)                   
   tidyverse   * 1.3.0      2019-11-21 [1] CRAN (R 4.0.2)                   
 V tinytex       0.24       2020-07-24 [1] CRAN (R 4.0.2)                   
   usethis       1.6.1      2020-04-29 [1] CRAN (R 4.0.2)                   
   vctrs         0.3.1      2020-06-05 [1] CRAN (R 4.0.2)                   
   whisker       0.4        2019-08-28 [1] CRAN (R 4.0.2)                   
   withr         2.2.0      2020-04-20 [1] CRAN (R 4.0.2)                   
   xfun          0.15       2020-06-21 [1] CRAN (R 4.0.2)                   
   xml2          1.3.2      2020-04-23 [1] CRAN (R 4.0.2)

**Comment by Torvaney** [comment]
Hmm, I am still running it fine with the same versions of broom and regista.

Do you see any output when you run `regista:::tidy.dixoncoles` (without brackets) in the console?

**Comment by minor-ls** [comment]
> regista:::tidy.dixoncoles
function (x, ...) 
{
    parameter_names <- strsplit(names(x$par), "___")
    parameter_values <- x$par
    pluck_na <- function(y, ...) {
        pluck(y, ...) %||% NA_character_
    }
    tibble::tibble(parameter = map_chr(parameter_names, pluck, 
        1), team = map_chr(parameter_names, pluck_na, 2), value = parameter_values)
}
<bytecode: 0x00000244795a4c30>
<environment: namespace:regista>


This link would suggest that it has something to do with the R session, but running it in a fresh session, as they suggested, didn't work either --> https://stackoverflow.com/questions/53728532/error-no-tidy-method-for-objects-of-class-lda-vem

**Comment by Torvaney** [comment]
How strange. Are you persisting any data between sessions? And is this an issue unique to R v4+ (I am still using 3.6.3)?

**Comment by minor-ls** [comment]
The second time I tried it, I used a brand new R session with no persisting data. I could try to roll back R to see if it works with an earlier version.

**Comment by Torvaney** [comment]
Hmmm, this sounds to me like it could be an issue with R itself and S3 methods rather than regista or broom.

Do you have the same issue using `augment` with anything else, for example `lm`?

**Comment by Torvaney** [comment]
Closing due to inactivity

**Comment by Torvaney** [comment]
Issue has been _properly_ fixed by 8f114d9a0a0bbfa2a3f85be4fe2b1ded1a0d1933
